README.md
# pm2-health
This [PM2](http://pm2.keymetrics.io/) module is:
* Monitoring events (like app exit, restart etc.)
* Monitoring app exceptions
* Monitoring PMX metrics of your apps and send alerts when value hits treshold
* Sending mail notifications

With rich config options you can fine-tune monitoring rules.

## Installation & Upgrade

`pm2 install pm2-health`

## Configuration

After installation run `pm2 conf` to configure module. Alternatively edit `module_conf.json` file directly (in PM2 home folder).

### Minimum required config:

```json
"pm2-health": {
    "smtp": {
        "host": "your-smtp-host",
        "port": 587,
        "from": "your-from-mail", // if not set, user will be used
        "user": "your-smtp-user",   // auth
        "password": "your-smtp-password",   // auth
        "secure": false,
        "disabled": false,
        "clientHostName": "your-machine.com", // optional, will force the client host-name FQDN used in SMTP HELLO. If not set, NodeMailer will ask host name to the OS, and use 127.0.0.1 if it's not a FQDN.
    },
    "mailTo": "mail1,mail2"
}
```
> If any of required properties is not defined, `pm2-health` will shutdown. You can check error logs for details.

### All config options:

* `smtp` - SMTP server configuration. If your SMTP doesn't require auth, leave `smtp.user` empty

* `mailTo` - comma separated list of notification receipients

* `replyTo` - reply to address (optional)

* `events` - list of events to monitor (optional). If not set, all events will be monitored. 

> Manually triggered events will not send notification.

* `exceptions` - if `true` apps exceptions will be monitored (optional)

* `messages` - if `true` apps custom messages will be monitored (optional). See [Custom messages](#custom-messages)

* `messageExcludeExps` - array of regular expressions used to exclude messages (optional). See [Filtering custom messages](#filtering-custom-messages)

* `metric` - object describing PMX metrics to be monitored (optional). See [Metrics monitoring](#metrics-monitoring)

* `metricIntervalS` - how often PMX metrics will be tested in seconds (optional). If not set, 60 seconds is used

* `aliveTimeoutS` - alive watchdog timeout interal in seconds. If not set watchdog function is off. See [Process alive watchdog](#process-alive-watchdog)

* `addLogs` - if `true` app logs will be added as mail attachement (optional)

* `appsExcluded` - array of app names to exclude from monitoring (optional)

* `appsIncluded` - array of app names to include, if set `appsExcluded` is ignored (optional)

* `webConfig` - if set, some of the config settings can be downloaded from given url (optional). See [Web config](#web-config)

* `debugLogEnabled` - if `true` debug log is enabled, by default is `false` (optional)

* `batchPeriodM` - enables message batching and sets batching period (optional). See [Message batching](#message-batching)

* `batchMaxMessages` - max. messages in batch (optional). See [Message batching](#message-batching)

## Metrics monitoring

`pm2-health` can monitor any PMX metrics defined in your apps.

To configure rules of alerting, setup `metric` section in module config file.

```json
"metric": {
    "metric name": {
        "target": 0,
        "op": ">",
        "ifChanged": true,
        "noNotify": true,
        "noHistory": true,
        "exclude": false
    },
    "metric 2": {
        ...
    }    
}
```
* `metric name` - name of metric defined in one of your apps

* `target` - target numeric value

* `op` - operator to compare metric value and target. Can be one of: `<`, `>`, `=`, `<=`, `>=`, `!=`

* `ifChanged` - if `true`, alert will trigger only if current metric value is different from last recorded value (optional)

* `noNotify` - if `true`, no alerts will be send (optional)

* `noHistory` - if `true`, metric value history won't be stored (optional)

* `exclude` - if `true`, metric will be complettely excluded from monitoring (optional)

* `direct` - if `true`, metric value won't be converted to number (optional)

By default, `cpu` and `memory` metrics are added.

> Learn how to define PMX metrics in your apps here: http://pm2.keymetrics.io/docs/usage/process-metrics/

## Custom messages

On top of standard PM2 events, you can monitor custom messages sent from your apps.

To send message from your app use:
```javascript
process.send({
    type: "process:msg",    
    data: {    
        ...
    }
});
```

* `type` - must be `process:msg`

* `data` - object containing additional data (optional).

### Filtering custom messages
You can exclude some of the messages based on their `data` content:

1. Add regular expression to list in `messageExcludeExp` config property
1. `data` (converted to JSON string) will be tested with this all expressions in the list
1. If any test is positive, message will be excluded

Example:

You wish to monitor slow operations in your app, so you send custom messages like so:
```javascript
function slow(operation, duration) {
    process.send({ type: "process:msg", data: { operation, duration }});
}
```
You know that `backup` and `restore` operations are always slow and wish to exclude them, but still get other slow operations.

Set config to:
```json
"messageExcludeExps": [
    "\"operation\": \"(backup|restore)\""
]
```
> Remember to escape regex string to JSON string

## Process alive watchdog

Alive watchdog (added in 1.9.0) can observe alive messages from processes.

To use functionallity your process has to send periodically `process:msg` signal as such:
```javascript
process.send({
    type: "process:msg",    
    data: "alive"
});
```

In addition config parameter `aliveTimeoutS` must be added. If alive message won't be received within `aliveTimeoutS` (seconds), alert will be send.

> `aliveTimeoutS` must be lower than interval of sending `process:alive` signal.

After first alert, following test is done every 10 minues for 6 consecutive times, after wich alerting stops, assuming process is permanetly closed.


## Web config

Web config (added in 1.7) allows you to fetch some of the config settings from web url.

Sample config:

```json
{
    "webConfig": {
        "url": "url of JSON file",
        "auth": {
            "user": "...",
            "password": "..."
        },
        "fetchIntervalM": 10
    }
}
```

Url must return UTF-8 JSON with config properties.

> Only following properties can be used: `events`, `metric`, `exceptions`, `messages`, `messageExcludeExps`, `appsExcluded`, `metricIntervalS`, `addLogs`, `batchPeriodM`, `batchMaxMessages`

## Message batching

Feature added in (1.11) allows to merge multiple messages over period of time and send them as single message.

This can be used to limit number of messages sent (prevent spam).

To enable please set following properties in config section:

- `batchPeriodM` - period (in minutes) to batch, if not set batching is not enabled
- `batchMaxMessages` - max. number of messages in batch (optional)

Batch message will be send after `batchPeriodM` elapses or if number of messages collected are greater than `batchMaxMessages`.

> Priority messages (as exceptions etc.) are not batched and sent immediatelly.

> It's advised to set `batchMaxMessages` to prevent huge messages.

> Batching settings can be changed by web config.

## Hold notifications temporarily

To hold mail notification: `pm2 trigger pm2-health hold 30`

> Notifications will restart automatically after 30 minutes.

To unhold immediatelly: `pm2 trigger pm2-health unheld`

> All monitoring processes continues, just mail notification is held

## Mail template

Mail uses HTML format. To adjust template, you can edit [Template.html](./Template.html)

`<!-- body -->` will be exchanged with actual message body.

`<!-- timeStamp -->` will be exchanged with event timestamp (UTC).

> `pm2-health` update will override your `Template.html`, so keep backup :blush:

## Testing mail

To send test mail: `pm2 trigger pm2-health mail`

## Building

`pm2-health` is written using TypeScript 2.6.1+ with `es2017` target. 

`es2017` is supported by Node 8+. If you need to use ealier version, build solution using `es5` or `es6` target.

> Solution includes VS Code settings for build and debug.

CHANGELOG.md
# 1.12.0
- added optional `clientHostName` into SMTP config

# 1.11.0
- added batching of messages

# 1.10.6
- revert pm2 version upgrade to pmx due pmx issues

# 1.10.0
- updated dependencies to newest
- added license file (#23)

# 1.9.5
- add `appsIncluded` to config (#18)

# 1.9.0
- add alive functionallity

# 1.8.9
- add important flag to some of the mails
- minnor fixes

# 1.8.0
- add app `inactive` flag for apps that don't sends signals anymore

# 1.7.3
- add `from` to `smtp` config (#16)

# 1.7.0
- add possibility of fetching config from url (#14)

# 1.6.0
- base

Health.ts
import * as PM2 from "pm2";
import * as Pmx from "pmx";
import * as Fs from "fs";
import { basename } from "path";
import { ISmtpConfig } from "./Mail";
import { Snapshot, IShapshotConfig, IAuth } from "./Snapshot";
import { Fetch } from "planck-http-fetch";
import { info, error, debug, enableDebugLog } from "./Log";
import { Notify } from "./Notify";

const
    MERTIC_INTERVAL_S = 60,
    HOLD_PERIOD_M = 30,
    ALIVE_MAX_CONSECUTIVE_TESTS = 6,
    ALIVE_CONSECUTIVE_TIMEOUT_S = 600,
    LOGS = ["pm_err_log_path", "pm_out_log_path"],
    OP = {
        "<": (a, b, t) => a < b && Math.abs(a - b) > t,
        ">": (a, b, t) => a > b && Math.abs(a - b) > t,
        "=": (a, b, t) => a === b,
        "~": (a, b, t) => Math.abs(a - b) > t,
        "<=": (a, b, t) => a <= b,
        ">=": (a, b, t) => a >= b,
        "!=": (a, b, t) => a !== b,
        "!~": (a, b, t) => Math.abs(a - b) > t
    };

interface IMonitConfig {
    events: string[];
    metric: {
        [key: string]: {
            target?: any;
            op?: "<" | ">" | "=" | "<=" | ">=" | "!=";
            ifChanged?: boolean;
            noHistory?: boolean;
            noNotify: boolean;
            exclude?: boolean;
            direct?: boolean;
            tolerance?: number;
        }
    },
    exceptions: boolean;
    messages: boolean;
    messageExcludeExps: string;
    appsIncluded: string[];
    appsExcluded: string[];
    metricIntervalS: number;
    addLogs: boolean;
    aliveTimeoutS: number;
}

interface IConfig extends IMonitConfig, ISmtpConfig, IShapshotConfig {
    webConfig: {
        url: string;
        auth?: IAuth;
        fetchIntervalM: number;
    },
    debugLogEnabled?: boolean;
}

// keys that can be updated by web config
const CONFIG_KEYS: (keyof IConfig)[] = ["events", "metric", "exceptions", "messages", "messageExcludeExps", "appsExcluded", "metricIntervalS", "addLogs", "aliveTimeoutS", "batchPeriodM", "batchMaxMessages", "mailTo", "replyTo"];

export class Health {
    readonly _notify: Notify;
    readonly _snapshot: Snapshot;

    constructor(private _config: IConfig) {
        if (this._config.debugLogEnabled === true)
            enableDebugLog();

        debug(JSON.stringify(this._config, undefined, 2));

        if (this._config.metricIntervalS == null || this._config.metricIntervalS < MERTIC_INTERVAL_S) {
            info(`setting default metric check interval ${MERTIC_INTERVAL_S} s.`);
            this._config.metricIntervalS = MERTIC_INTERVAL_S;
        }

        if (!this._config.metric)
            this._config.metric = {};

        this._notify = new Notify(_config);
        this._snapshot = new Snapshot(this._config);
    }

    async fetchConfig() {
        // todo: what if web config contains invalid data, changes should be reversed
        try {
            info(`fetching config from [${this._config.webConfig.url}]`);

            const fetch = new Fetch(this._config.webConfig.url);

            if (this._config.webConfig.auth && this._config.webConfig.auth.user)  // auth
                fetch.basicAuth(this._config.webConfig.auth.user, this._config.webConfig.auth.password);

            const
                json = await fetch.fetch(),
                config = JSON.parse(json);

            // map config keys
            for (const key of CONFIG_KEYS)
                if (config[key] != null) {
                    this._config[<string>key] = config[key];
                    info(`applying [${key}] = ${config[key]}`);
                }

            this.configChanged();
        }
        catch (ex) {
            error(`failed to fetch config -> ${ex.message || ex}`);
        }
    }

    _messageExcludeExps: RegExp[];

    configChanged() {
        this._messageExcludeExps = [];
        if (Array.isArray(this._config.messageExcludeExps))
            this._messageExcludeExps = this._config.messageExcludeExps.map(e => new RegExp(e));

        this._notify.configChanged();
    }

    isAppIncluded(app: string) {
        if (app === "pm2-health")
            return false;

        if (Array.isArray(this._config.appsIncluded))
            return this._config.appsIncluded.includes(app);

        if (Array.isArray(this._config.appsExcluded))
            return !this._config.appsExcluded.includes(app);

        return false;
    }

    async go() {
        info(`pm2-health is on`);

        this.configChanged();

        // fetch web config (if set)
        if (this._config.webConfig && this._config.webConfig.url) {
            await this.fetchConfig();

            if (this._config.webConfig.fetchIntervalM > 0)
                setInterval(
                    () => {
                        this.fetchConfig();
                    },
                    this._config.webConfig.fetchIntervalM * 60 * 1000);
        }

        PM2.connect((ex) => {
            stopIfEx(ex);

            PM2.launchBus((ex, bus) => {
                stopIfEx(ex);

                bus.on("process:event", (data) => {
                    if (data.manually || !this.isAppIncluded(data.process.name))
                        return;

                    if (Array.isArray(this._config.events) && this._config.events.indexOf(data.event) === -1)
                        return;

                    this._notify.send({
                        subject: `${data.process.name}:${data.process.pm_id} - ${data.event}`,
                        body: `
                        <p>App: <b>${data.process.name}:${data.process.pm_id}</b></p>
                        <p>Event: <b>${data.event}</b></p>
                        <pre>${JSON.stringify(data, undefined, 4)}</pre>`,
                        priority: "high",
                        attachements: LOGS.filter(e => this._config.addLogs === true && data.process[e]).map(e => ({ filename: basename(data.process[e]), path: data.process[e] }))
                    });
                });

                if (this._config.exceptions)
                    bus.on("process:exception", (data) => {
                        if (!this.isAppIncluded(data.process.name))
                            return;

                        this._notify.send({
                            subject: `${data.process.name}:${data.process.pm_id} - exception`,
                            body: `
                            <p>App: <b>${data.process.name}:${data.process.pm_id}</b></p>                            
                            <pre>${JSON.stringify(data.data, undefined, 4)}</pre>`,
                            priority: "high"
                        });
                    });

                if (this._config.messages)
                    bus.on("process:msg", (data) => {
                        if (!this.isAppIncluded(data.process.name))
                            return;

                        if (data.data === "alive") {
                            this.aliveReset(data.process, this._config.aliveTimeoutS);
                            return;
                        }

                        const json = JSON.stringify(data.data, undefined, 4);

                        if (this._messageExcludeExps.some(e => e.test(json)))
                            return; // exclude

                        this._notify.send({
                            subject: `${data.process.name}:${data.process.pm_id} - message`,
                            body: `
                            <p>App: <b>${data.process.name}:${data.process.pm_id}</b></p>
                            <pre>${json}</pre>`
                        });
                    });
            });

            this.testProbes();
        });

        Pmx.action("hold", undefined, (p, reply) => {
            let t = HOLD_PERIOD_M;
            if (p) {
                const n = Number.parseInt(p);
                if (!Number.isNaN(n))
                    t = n;
            }

            const holdTill = new Date();
            holdTill.setTime(holdTill.getTime() + t * 60000);

            this._notify.hold(holdTill)

            const msg = `mail held for ${t} minutes, till ${holdTill.toISOString()}`;
            info(msg);
            reply(msg);
        });

        Pmx.action("unheld", undefined, (reply) => {
            this._notify.hold(null);
            info("mail unheld");
            reply("mail unheld");
        });

        Pmx.action("mail", undefined, async (reply) => {
            try {
                await this._notify.send({ subject: "Test only", body: "This is test only.", priority: "high" });    // high -> to bypass batching
                info("mail send");
                reply("mail send");
            }
            catch (ex) {
                reply(`mail failed: ${ex.message || ex}`);
            }
        });

        Pmx.action("dump", undefined, (reply) => {
            this._snapshot.dump();
            reply(`dumping`);
        });

        // for dev. only
        Pmx.action("debug", undefined, (reply) => {
            PM2.list((ex, list) => {
                stopIfEx(ex);

                Fs.writeFileSync("pm2-health-debug.json", JSON.stringify(list));
                Fs.writeFileSync("pm2-health-config.json", JSON.stringify(this._config));

                reply(`dumping`);
            });
        });
    }

    private _timeouts = new Map<string, NodeJS.Timer>();

    private aliveReset(process: { name, pm_id }, timeoutS: number, count = 1) {
        clearTimeout(this._timeouts.get(process.name));

        this._timeouts.set(
            process.name,
            setTimeout(() => {
                info(`death ${process.name}:${process.pm_id}, count ${count}`);

                this._notify.send({
                    subject: `${process.name}:${process.pm_id} - is death!`,
                    body: `
                    <p>App: <b>${process.name}:${process.pm_id}</b></p>
                    <p>This is <b>${count}/${ALIVE_MAX_CONSECUTIVE_TESTS}</b> consecutive notice.</p>`,
                    priority: "high"
                });

                if (count < ALIVE_MAX_CONSECUTIVE_TESTS)
                    this.aliveReset(process, ALIVE_CONSECUTIVE_TIMEOUT_S, count + 1);
            }, timeoutS * 1000));
    }

    private testProbes() {
        debug("testing probes");

        const alerts = [];

        PM2.list(async (ex, list) => {
            stopIfEx(ex);

            for (const app of list) {
                if (!this.isAppIncluded(app.name))
                    continue;

                let monit = app.pm2_env["axm_monitor"];
                if (!monit)
                    monit = {};

                // add memory + cpu metrics
                if (app.monit) {
                    monit["memory"] = { value: app.monit.memory / 1048576 };
                    monit["cpu"] = { value: app.monit.cpu };
                }

                if (app.pm2_env) {
                    if (app.pm2_env["_pm2_version"])
                        monit["pm2"] = { value: app.pm2_env["_pm2_version"], direct: true };

                    if (app.pm2_env["node_version"])
                        monit["node"] = { value: app.pm2_env["node_version"], direct: true };
                }

                for (const key of Object.keys(monit)) {
                    let probe = this._config.metric[key];
                    if (!probe)
                        probe = { noNotify: true, direct: monit[key].direct === true, noHistory: monit[key].direct === true };

                    if (probe.exclude === true)
                        continue;

                    let
                        v = monit[key].value,
                        bad: boolean;

                    if (!probe.direct) {
                        v = Number.parseFloat(v);
                        if (Number.isNaN(v)) {
                            error(`monit [${app.name}.${key}] -> [${monit[key].value}] is not a number`);
                            continue;
                        }
                    }

                    if (probe.op && probe.op in OP && probe.target != null)
                        bad = OP[probe.op](v, probe.target, probe.tolerance || 0);

                    // test
                    if (probe.noNotify !== true && bad === true && (probe.ifChanged !== true || this._snapshot.last(app.pm_id, key) !== v))
                        alerts.push(`<tr><td>${app.name}:${app.pm_id}</td><td>${key}</td><td>${v}</td><td>${this._snapshot.last(app.pm_id, key)}</td><td>${probe.target}</td></tr>`);

                    const data: any = { v }
                    if (bad)    // safe space by not storing false
                        data.bad = true;

                    this._snapshot.push(app.pm_id, app.name, key, !probe.noHistory, data);
                }
            }

            this._snapshot.inactivate();
            await this._snapshot.send();

            if (alerts.length > 0)
                this._notify.send({
                    subject: `${alerts.length} alert(s)`,
                    body: `
                    <table>
                        <tr>
                            <th>App</th><th>Metric</th><th>Value</th><th>Prev. Value</th><th>Target</th>
                        </tr>
                        ${alerts.join("")}
                    </table>`,
                    priority: "high"
                });

            setTimeout(
                () => { this.testProbes(); },
                1000 * this._config.metricIntervalS);
        });
    }
}

export function stopIfEx(ex: any) {
    if (ex) {
        error(ex.message || ex);
        PM2.disconnect();
        process.exit(1);
    }
}